### PR TITLE
Update `SortedMerge` to use a `PriorityQueue`

### LINKED ISSUE #654: `PriorityQueue` implementation of `SortedMergeBy`
I want to do typical external merge sort over a large-ish number of sorted buckets, and `SortedMerge` gets me the fully sorted results after I presort chunks. For large data sets I have enough memory to do this across a few hundred sorted chunks.

Problem is that `SortedMerge` gets slower linearly with the number of enumerables it merges across, because it does a linear search through the remaining enumerators, performing key comparisons with half of them on average.

An alternative is to use the CLR's `PriorityQueue` to store the enumerators, which is very fast at returning the item with the lowest key, so the key here is just the comparison key obtained the from enumerators' `Current`  property.

Implementation and test results (done on a release build) here:

https://github.com/danielearwicker/PriorityQueueMergeSort

Rough idea:

```cs
var queue = new PriorityQueue<IEnumerator<TSource>, TKey>(comparer);

foreach (var e in enumerators)
{
    if (e.MoveNext()) queue.Enqueue(e, keySelector(e.Current));
}

while (queue.TryDequeue(out var e, out var _))
{
    yield return e.Current;

    if (e.MoveNext()) queue.Enqueue(e, keySelector(e.Current));
}
```

Time to sort `int`s split randomly between N sorted collections, N = 2 to 30:

image

The current implementation is $O(N)$, the PQ version is $O(\log N)$. At N = 100, PQ is over 8x faster. Sorting `string`s it's 6x faster.

But:

1. For very small N it's slightly slower, the overhead of delete/insert on the PQ overpowers the saving on comparisons.
2. `PriorityQueue` makes no guarantees about the ordering of equal keys based on how they were inserted. So the sort may behave differently where keys are equal and that might theoretically affect existing apps.
3. If you have two lists to merge then the `left.SortedMerge(right)` pattern is natural. If you have 100, `sources.First().SortedMerge(sources.Skip(1))` is awkward.

So given 1 and 2 it seems safer to let the user choose the implementation. Seems like one natural way to do this would be to solve 3 at the same time and expose it as an extension method on `IEnumerable<IEnumerable<T>>`, so if you have a sequence of sorted sequences (as you would if there were lots of them), you can directly call `SortedMerge` on it, and the implementation will be suitable for many sequences.

I'd be happy to PR this, or another approach?

### PR DESCRIPTION
**Pull request by danielearwicker**
Fixes #654

### CODE REVIEW

File: Source/SuperLinq/SortedMerge.cs [hunk 1]
```diff
@@ -355,48 +355,124 @@ public static IEnumerable<TSource> SortedMerge<TSource>(this IEnumerable<TSource
...
+#if NET6_0_OR_GREATER
+				var queue = new PriorityQueue<IEnumerator<TSource>, TKey>(
+					enumerators.Select(x => (x, keySelector(x.Current))),
+					comparer);
+
+#pragma warning disable CA2000 // e will be disposed via enumerators list
+				while (queue.TryDequeue(out var e, out var _))
+#pragma warning restore CA2000 // Dispose objects before losing scope
+				{
+					yield return e.Current;
```
Comment (viceroypenguin) [inline Source/SuperLinq/SortedMerge.cs:392]:
```suggestion
				while (queue.TryDequeue(out var e, out var current))
#pragma warning restore CA2000 // Dispose objects before losing scope
				{
					yield return current;
```

File: Source/SuperLinq/SortedMerge.cs [hunk 2]
```diff
@@ -355,48 +355,124 @@ public static IEnumerable<TSource> SortedMerge<TSource>(this IEnumerable<TSource
...
+
+#if !NET6_0_OR_GREATER
+	internal sealed record class SourceComparer<TItem, TKey>(
+		IComparer<TKey> KeyComparer,
+		Func<TItem, TKey> KeySelector
+	) : IComparer<IEnumerator<TItem>>
+	{
+		public int Compare(IEnumerator<TItem>? x, IEnumerator<TItem>? y)
+			=> KeyComparer.Compare(KeySelector(x!.Current), KeySelector(y!.Current));
+	}
```
Comment (viceroypenguin) [inline Source/SuperLinq/SortedMerge.cs:476]:
```suggestion
	internal sealed class SourceComparer<TItem, TKey>(
		IComparer<TKey> keyComparer,
		Func<TItem, TKey> keySelector
	) : IComparer<IEnumerator<TItem>>
	{
		public int Compare(IEnumerator<TItem>? x, IEnumerator<TItem>? y)
			=> keyComparer.Compare(keySelector(x!.Current), keySelector(y!.Current));
	}
```

File: Source/SuperLinq/SortedMerge.cs [hunk 3]
```diff
@@ -355,48 +355,124 @@ public static IEnumerable<TSource> SortedMerge<TSource>(this IEnumerable<TSource
...
+				{
+					var e = sequence.GetEnumerator();
+					if (e.MoveNext())
+					{
+						enumerators.Add(e);
+					}
+					else
+					{
+						e.Dispose();
+					}
```
Comment (viceroypenguin) [inline Source/SuperLinq/SortedMerge.cs:380]:
```suggestion
					if (e.MoveNext())
						enumerators.Add(e);
					else
						e.Dispose();
```

File: Source/SuperLinq/SortedMerge.cs [hunk 4]
```diff
@@ -355,48 +355,124 @@ public static IEnumerable<TSource> SortedMerge<TSource>(this IEnumerable<TSource
...
+							yield return e.Current;
+						}
+
+						break;
+					}
+
+					if (e.MoveNext())
+					{
+						queue.Enqueue(e, keySelector(e.Current));
+					}
```
Comment (viceroypenguin) [inline Source/SuperLinq/SortedMerge.cs:408]:
```suggestion
					if (e.MoveNext())
						queue.Enqueue(e, keySelector(e.Current));
```

**Review by viceroypenguin — COMMENTED** [review]
comments are mostly style nits - if you could fix other single line blocks to not use `{}`  as well, that would be good. also, there's a build error, if you could look at that please.

File: Source/SuperLinq/SortedMerge.cs (hunk 1 above)
Comment (danielearwicker) [reply Source/SuperLinq/SortedMerge.cs]: The second output isn't `e.Current`, it's the key, i.e.`keySelector(e.Current)`.

File: Source/SuperLinq.Async/SortedMerge.cs [hunk 5]
```diff
@@ -291,39 +288,91 @@ public static partial class AsyncSuperEnumerable
...
+			}
+			finally
+			{
+				foreach (var e in enumerators) await e.DisposeAsync();
 			}
 		}
 	}
+
+#if !NET6_0_OR_GREATER
+	internal sealed record class SourceComparer<TItem, TKey>(
```
Comment (viceroypenguin) [inline Source/SuperLinq.Async/SortedMerge.cs:369]: The style guideline failed because this is still a `record class`, when it should be a `class`.

**Comment by danielearwicker** [comment]
Looks like Test.Async.EndsWithTest.EndsWithWithStrings has some intermittent timing issue?

**Comment by viceroypenguin** [comment]
Thanks!

**Pull request merged**